### PR TITLE
Refactor server to use the daemon as a service

### PR DESCRIPTION
**Pull request by runcom**
Follow up of #12937
This pr clean server.go to deal with container struct. Everything that should be done on a container should be handled from the daemon (acting as the service layer). Adapting types, new api version, upgrading should be a breeze now since no logic nor nothing is handled in controllers anymore.
I tried to keep AttachWithLogs handler as simple as possible to deal with container configuration and version like ContainersLogs is already doing.
/cc @tiborvass since you looked at the previous pr
@calavera for api too

Signed-off-by: Antonio Murdaca me@runcom.ninja

### CODE REVIEW

File: api/server/server.go [hunk 1]
```diff
@@ -1062,14 +1062,11 @@ func (s *Server) postContainersWait(version version.Version, w http.ResponseWrit
 		return fmt.Errorf("Missing parameter")
 	}
 
-	name := vars["name"]
-	cont, err := s.daemon.Get(name)
-	if err != nil {
+	status, err := s.daemon.ContainerWait(vars["name"], -1*time.Second)
+	if err != nil && !strings.Contains(err.Error(), "Timed out") {
```
Comment (runcom) [inline api/server/server.go]: This error only happens if container.WaitStop time out. Could make sense to create a new typed error in daemon/container and type checking it here instead of going with strings matching?

Comment (calavera) [inline api/server/server.go]: yes! the less string checks the better :+1:

Comment (runcom) [inline api/server/server.go]: totally

Comment (runcom) [inline api/server/server.go]: ok forget this, typing the error as state.SomeWaitError will import here in api/server/server.go and I personally don't like it

Comment (cpuguy83) [inline api/server/server.go]: Could make an errors subpackage for state.... or just a generic errors subpackage on daemon.
Testing error values is bad.

Comment (cpuguy83) [inline api/server/server.go]: Also can just be for a separate PR for that.

Comment (runcom) [inline api/server/server.go]: Yeah I agree with you testing error values is bad, I'll carry this in another pr

File: api/server/server.go [hunk 2]
```diff
@@ -1103,50 +1100,33 @@ func (s *Server) postContainersAttach(version version.Version, w http.ResponseWr
...
-		errStream = outStream
+	attachWithLogsConfig := &daemon.ContainerAttachWithLogsConfig{
+		InStream:   inStream,
+		OutStream:  outStream,
+		UseStdin:   boolValue(r, "stdin"),
+		UseStdout:  boolValue(r, "stdout"),
+		UseStderr:  boolValue(r, "stderr"),
+		Logs:       boolValue(r, "logs"),
+		Stream:     boolValue(r, "stream"),
+		ApiVersion: version,
```
Comment (vieux) [inline api/server/server.go]: I don't know about the others, but I'm personally not fond of leaking the ApiVersion to other pkg that the API

Comment (runcom) [inline api/server/server.go]: @vieux this was one of my concern as well but since the stream configuration depends also on apiversion I couldn't do other way (we did the same for api logs call)

**Comment by runcom** [comment]
ping @vieux updated (forget my previous comment)

**Comment by runcom** [comment]
ping @calavera @tiborvass

File: api/server/server.go [hunk 3]
```diff
@@ -1058,14 +1058,11 @@ func (s *Server) postContainersWait(version version.Version, w http.ResponseWrit
 		return fmt.Errorf("Missing parameter")
 	}
 
-	name := vars["name"]
-	cont, err := s.daemon.Get(name)
-	if err != nil {
+	status, err := s.daemon.ContainerWait(vars["name"], -1*time.Second)
+	if err != nil && !strings.Contains(strings.ToLower(err.Error()), "timed out") {
```
Comment (duglin) [inline api/server/server.go]: with a timeout of `-1` would we ever see a "timed out" error?  I'm not sure we would.

Comment (duglin) [inline api/server/server.go]: But even if we did, I'm not sure we'd want to return a `200` on  a time out. That would be misleading to the user.
I think we should just remove the && part of the previous `if`

File: api/server/server.go [hunk 4]
```diff
@@ -1099,50 +1096,33 @@ func (s *Server) postContainersAttach(version version.Version, w http.ResponseWr
...
 	}
 
-	if !cont.Config.Tty && version.GreaterThanOrEqualTo("1.6") {
-		errStream = stdcopy.NewStdWriter(outStream, stdcopy.Stderr)
-		outStream = stdcopy.NewStdWriter(outStream, stdcopy.Stdout)
-	} else {
-		errStream = outStream
+	attachWithLogsConfig := &daemon.ContainerAttachWithLogsConfig{
+		InStream:  inStream,
+		OutStream: outStream,
```
Comment (duglin) [inline api/server/server.go]: Why don't we need to pass in errStream?

Comment (duglin) [inline api/server/server.go]: Never mind - I see you moved it into the daemon code.

File: api/server/server.go (hunk 3 above)
Comment (runcom) [inline api/server/server.go]: I'm ok with what you said, we are ignoring the error, I don't know if there was a reason, maybe @LK4D4 ?
this is old code:
```
status, _ := cont.WaitStop(-1 * time.Second)
```

Comment (duglin) [inline api/server/server.go]: Unless someone can come up with a good reason I wouldn't ignore the error, let's just return it.  I think the only error we could get in this case is the "container not found" one, which I think is valid to return.

Comment (runcom) [inline api/server/server.go]: alright

**Comment by runcom** [comment]
@duglin updated

**Comment by tiborvass** [comment]
As much as I welcome this PR overall, I still think we should make sure we are consistent.

Currently there are methods on containers to do things, and this PR would introduce methods on the daemon to manipulate the containers by name.

I know for sure that @crosbymichael wanted to make containers just structs without methods, and have the daemon take container instances in its methods, which I think would be really nice. In that light, should we suffix the daemon methods in this PR with "ByName" or do that in due time?

**Comment by LK4D4** [comment]
I don't want to introduce suffixes in favor of currently non-existent concepts. And I don't understand from where we will take instances to pass them to daemon.

**Comment by tiborvass** [comment]
@LK4D4 fair enough :)

LGTM

**Comment by runcom** [comment]
ping @duglin could you take another look?

File: api/server/server.go [hunk 5]
```diff
@@ -1099,50 +1096,33 @@ func (s *Server) postContainersAttach(version version.Version, w http.ResponseWr
...
-	}
-	if boolValue(r, "stdout") {
-		stdout = outStream
-	}
-	if boolValue(r, "stderr") {
-		stderr = errStream
+	attachWithLogsConfig := &daemon.ContainerAttachWithLogsConfig{
+		InStream:  inStream,
+		OutStream: outStream,
+		UseStdin:  boolValue(r, "stdin"),
```
Comment (duglin) [inline api/server/server.go]: Not a big deal, but is UseStdin is false, then InStream is ignored later on, right? If so, could we get the same net result if we just set InStream to nil?  Same for OutStream?

**Comment by duglin** [comment]
just one question... other than that LGTM

File: api/server/server.go (hunk 5 above)
Comment (runcom) [inline api/server/server.go]: yup but I preferred to move the whole logic inside the daemon function instead of checking what you say here in the controller

**Comment by LK4D4** [comment]
LGTM
